The errors are very weird, the build system seems to have used the original version of DataSampling.cxx instead of the one from this PR:
```
/System/Volumes/Data/build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/4605/0/Utilities/DataSampling/src/DataSampling.cxx:137:26: error: no matching constructor for initialization of 'o2::utilities::DataSamplingPolicy'
      DataSamplingPolicy policy(policyConfig.second);
```
I will try to force-push and see if that happens again.

--- a/Utilities/DataSampling/include/DataSampling/DataSamplingCondition.h
+++ b/Utilities/DataSampling/include/DataSampling/DataSamplingCondition.h
@@ -24,7 +24,7 @@
 namespace o2::utilities
 {
 
-/// A standarised data sampling condition, to decide if given data sample should be passed forward.
+/// A standardised data sampling condition, to decide if given data sample should be passed forward.
 class DataSamplingCondition
 {
  public:

--- a/Utilities/DataSampling/include/DataSampling/DataSamplingPolicy.h
+++ b/Utilities/DataSampling/include/DataSampling/DataSamplingPolicy.h
@@ -39,8 +39,8 @@ class DataSamplingPolicy
   class PathMap : public PathVectorBase
   {
    public:
-    ~PathMap() = default;
     PathMap() = default;
+    ~PathMap() = default;
     const PathVectorBase::const_iterator find(const framework::ConcreteDataMatcher& input) const
     {
       return std::find_if(begin(), end(), [input](const auto& el) {
@@ -50,21 +50,33 @@ class DataSamplingPolicy
   };
 
  public:
-  /// \brief Constructor.
-  DataSamplingPolicy();
-  /// \brief Constructor.
-  DataSamplingPolicy(const boost::property_tree::ptree&);
-  /// \brief Destructor
-  ~DataSamplingPolicy();
-
   /// \brief Configures a policy using structured configuration entry.
-  void configure(const boost::property_tree::ptree&);
+  static DataSamplingPolicy fromConfiguration(const boost::property_tree::ptree&);
+
+  /// \brief Constructor.
+  DataSamplingPolicy(std::string name);
+  /// \brief Destructor
+  ~DataSamplingPolicy() = default;
+  /// \brief Move constructor
+  DataSamplingPolicy(DataSamplingPolicy&&) = default;
+  /// \brief Move assignment
+  DataSamplingPolicy& operator=(DataSamplingPolicy&&) = default;
+
+  /// \brief Adds a new association between inputs and outputs.
+  void registerPath(const framework::InputSpec&, const framework::OutputSpec&);
+  /// \brief Adds a new association between inputs and outputs.
+  //  void registerPolicy(framework::InputSpec&&, framework::OutputSpec&&);
+  /// \brief Adds a new sampling condition.
+  void registerCondition(std::unique_ptr<DataSamplingCondition>&&);
+  /// \brief Sets a raw FairMQChannel. Deprecated, do not use.
+  void setFairMQOutputChannel(std::string);
+
   /// \brief Returns true if this policy requires data with given InputSpec.
   bool match(const framework::ConcreteDataMatcher& input) const;
   /// \brief Returns true if user-defined conditions of sampling are fulfilled.
   bool decide(const o2::framework::DataRef&);
   /// \brief Returns Output for given InputSpec to pass data forward.
-  const framework::Output prepareOutput(const framework::ConcreteDataMatcher& input, framework::Lifetime lifetime = framework::Lifetime::Timeframe) const;
+  framework::Output prepareOutput(const framework::ConcreteDataMatcher& input, framework::Lifetime lifetime = framework::Lifetime::Timeframe) const;
 
   const std::string& getName() const;
   const PathMap& getPathMap() const;

--- a/Utilities/DataSampling/include/DataSampling/Dispatcher.h
+++ b/Utilities/DataSampling/include/DataSampling/Dispatcher.h
@@ -50,12 +50,14 @@ class Dispatcher : public framework::Task
   /// \brief Dispatcher process callback
   void run(framework::ProcessingContext& ctx) override;
 
-  /// \brief Create appropriate inputSpecs and outputSpecs for sampled data during the workflow declaration phase.
-  void registerPath(const std::pair<framework::InputSpec, framework::OutputSpec>&);
+  /// \brief Register a Data Sampling Policy
+  void registerPolicy(std::unique_ptr<DataSamplingPolicy>&&);
 
   const std::string& getName();
+  /// \brief Assembles InputSpecs of all registered policies in a single vector, removing overlapping entries.
   framework::Inputs getInputSpecs();
   framework::Outputs getOutputSpecs();
+  framework::Options getOptions();
 
  private:
   DataSamplingHeader prepareDataSamplingHeader(const DataSamplingPolicy& policy, const framework::DeviceSpec& spec);
@@ -67,8 +69,6 @@ class Dispatcher : public framework::Task
 
   std::string mName;
   std::string mReconfigurationSource;
-  framework::Inputs inputs;
-  framework::Outputs outputs;
   // policies should be shared between all pipeline threads
   std::vector<std::shared_ptr<DataSamplingPolicy>> mPolicies;
 };

--- a/Utilities/DataSampling/src/DataSampling.cxx
+++ b/Utilities/DataSampling/src/DataSampling.cxx
@@ -56,7 +56,6 @@ void DataSampling::GenerateInfrastructure(WorkflowSpec& workflow, const boost::p
 void DataSampling::DoGenerateInfrastructure(Dispatcher& dispatcher, WorkflowSpec& workflow, const boost::property_tree::ptree& policiesTree, size_t threads)
 {
   LOG(DEBUG) << "Generating Data Sampling infrastructure...";
-  Options options;
 
   for (auto&& policyConfig : policiesTree) {
 
@@ -64,7 +63,7 @@ void DataSampling::DoGenerateInfrastructure(Dispatcher& dispatcher, WorkflowSpec
 
     // We don't want the Dispatcher to exit due to one faulty Policy
     try {
-      policy = std::make_unique<DataSamplingPolicy>(policyConfig.second);
+      dispatcher.registerPolicy(std::make_unique<DataSamplingPolicy>(DataSamplingPolicy::fromConfiguration(policyConfig.second)));
     } catch (const std::exception& ex) {
       LOG(WARN) << "Could not load the Data Sampling Policy '"
                 << policyConfig.second.get_optional<std::string>("id").value_or("") << "', because: " << ex.what();
@@ -74,27 +73,17 @@ void DataSampling::DoGenerateInfrastructure(Dispatcher& dispatcher, WorkflowSpec
                 << policyConfig.second.get_optional<std::string>("id").value_or("") << "'";
       continue;
     }
-
-    for (const auto& path : policy->getPathMap()) {
-      dispatcher.registerPath({path.first, path.second});
-    }
-
-    if (!policy->getFairMQOutputChannel().empty()) {
-      options.push_back({"channel-config", VariantType::String, policy->getFairMQOutputChannel().c_str(), {"Out-of-band channel config"}});
-      LOG(DEBUG) << " - registering output FairMQ channel '" << policy->getFairMQOutputChannel() << "'";
-    }
   }
-  options.push_back({"period-timer-stats", framework::VariantType::Int, 10 * 1000000, {"Dispatcher's stats timer period"}});
 
   if (dispatcher.getInputSpecs().size() > 0) {
     DataProcessorSpec spec;
     spec.name = dispatcher.getName();
     spec.inputs = dispatcher.getInputSpecs();
     spec.outputs = dispatcher.getOutputSpecs();
-    spec.algorithm = adaptFromTask<Dispatcher>(std::move(dispatcher));
     spec.maxInputTimeslices = threads;
     spec.labels = {{"DataSampling"}, {"Dispatcher"}};
-    spec.options = options;
+    spec.options = dispatcher.getOptions();
+    spec.algorithm = adaptFromTask<Dispatcher>(std::move(dispatcher));
 
     workflow.emplace_back(std::move(spec));
   } else {
@@ -127,7 +116,7 @@ std::vector<InputSpec> DataSampling::InputSpecsForPolicy(ConfigurationInterface*
 
   for (auto&& policyConfig : policiesTree) {
     if (policyConfig.second.get<std::string>("id") == policyName) {
-      DataSamplingPolicy policy(policyConfig.second);
+      auto policy = DataSamplingPolicy::fromConfiguration(policyConfig.second);
       for (const auto& path : policy.getPathMap()) {
         InputSpec input = DataSpecUtils::matchingInput(path.second);
         inputs.push_back(input);
@@ -169,7 +158,7 @@ std::vector<OutputSpec> DataSampling::OutputSpecsForPolicy(ConfigurationInterfac
 
   for (auto&& policyConfig : policiesTree) {
     if (policyConfig.second.get<std::string>("id") == policyName) {
-      DataSamplingPolicy policy(policyConfig.second);
+      auto policy = DataSamplingPolicy::fromConfiguration(policyConfig.second);
       for (const auto& path : policy.getPathMap()) {
         outputs.push_back(path.second);
       }

--- a/Utilities/DataSampling/src/DataSampling.cxx
+++ b/Utilities/DataSampling/src/DataSampling.cxx
@@ -134,7 +134,7 @@ std::vector<InputSpec> DataSampling::InputSpecsForPolicy(std::shared_ptr<configu
 
   for (auto&& policyConfig : policiesTree) {
     if (policyConfig.second.get<std::string>("id") == policyName) {
-      DataSamplingPolicy policy(policyConfig.second);
+      auto policy = DataSamplingPolicy::fromConfiguration(policyConfig.second);
       for (const auto& path : policy.getPathMap()) {
         InputSpec input = DataSpecUtils::matchingInput(path.second);
         inputs.push_back(input);

--- a/Utilities/DataSampling/src/DataSamplingPolicy.cxx
+++ b/Utilities/DataSampling/src/DataSamplingPolicy.cxx
@@ -29,28 +29,30 @@ namespace o2::utilities
 
 using boost::property_tree::ptree;
 
-DataSamplingPolicy::DataSamplingPolicy() = default;
-
-DataSamplingPolicy::DataSamplingPolicy(const ptree& config)
+DataSamplingPolicy::DataSamplingPolicy(std::string name) : mName(std::move(name))
 {
-  configure(config);
 }
 
-DataSamplingPolicy::~DataSamplingPolicy() = default;
-
-void DataSamplingPolicy::configure(const ptree& config)
+void DataSamplingPolicy::registerPath(const InputSpec& inputSpec, const OutputSpec& outputSpec)
 {
-  mName = config.get<std::string>("id");
-  if (mName.size() > 14) {
-    LOG(WARNING) << "DataSamplingPolicy name '" << mName << "' is longer than 14 characters, we have to trim it. "
-                 << "Use a shorter policy name to avoid potential output name conflicts.";
-    mName.resize(14);
-  }
+  mPaths.emplace_back(inputSpec, outputSpec);
+}
 
-  auto subSpecString = config.get_optional<std::string>("subSpec").value_or("*");
-  auto subSpec = subSpecString.find_first_of("-*") != std::string::npos ? -1 : std::strtoull(subSpecString.c_str(), nullptr, 10);
+void DataSamplingPolicy::registerCondition(std::unique_ptr<DataSamplingCondition>&& condition)
+{
+  mConditions.emplace_back(std::move(condition));
+}
 
-  mPaths.clear();
+void DataSamplingPolicy::setFairMQOutputChannel(std::string channel)
+{
+  mFairMQOutputChannel = std::move(channel);
+}
+
+DataSamplingPolicy DataSamplingPolicy::fromConfiguration(const ptree& config)
+{
+  auto name = config.get<std::string>("id");
+  DataSamplingPolicy policy(name);
+
   size_t outputId = 0;
   std::vector<InputSpec> inputSpecs = DataDescriptorQueryBuilder::parse(config.get<std::string>("query").c_str());
 
@@ -60,29 +62,31 @@ void DataSamplingPolicy::configure(const ptree& config)
       OutputSpec outputSpec{
         {inputSpec.binding},
         createPolicyDataOrigin(),
-        createPolicyDataDescription(mName, outputId++),
+        createPolicyDataDescription(name, outputId++),
         DataSpecUtils::getOptionalSubSpec(inputSpec).value(),
         inputSpec.lifetime};
 
-      mPaths.push_back({inputSpec, outputSpec});
+      policy.registerPath(inputSpec, outputSpec);
 
     } else {
       OutputSpec outputSpec{
         {inputSpec.binding},
-        {createPolicyDataOrigin(), createPolicyDataDescription(mName, outputId++)},
+        {createPolicyDataOrigin(), createPolicyDataDescription(name, outputId++)},
         inputSpec.lifetime};
 
-      mPaths.push_back({inputSpec, outputSpec});
+      policy.registerPath(inputSpec, outputSpec);
     }
   }
 
-  mConditions.clear();
   for (const auto& conditionConfig : config.get_child("samplingConditions")) {
-    mConditions.push_back(DataSamplingConditionFactory::create(conditionConfig.second.get<std::string>("condition")));
-    mConditions.back()->configure(conditionConfig.second);
+    auto condition = DataSamplingConditionFactory::create(conditionConfig.second.get<std::string>("condition"));
+    condition->configure(conditionConfig.second);
+    policy.registerCondition(std::move(condition));
   }
 
-  mFairMQOutputChannel = config.get_optional<std::string>("fairMQOutput").value_or("");
+  policy.setFairMQOutputChannel(config.get_optional<std::string>("fairMQOutput").value_or(""));
+
+  return policy;
 }
 
 bool DataSamplingPolicy::match(const ConcreteDataMatcher& input) const
@@ -103,7 +107,7 @@ bool DataSamplingPolicy::decide(const o2::framework::DataRef& dataRef)
   return decision;
 }
 
-const Output DataSamplingPolicy::prepareOutput(const ConcreteDataMatcher& input, Lifetime lifetime) const
+Output DataSamplingPolicy::prepareOutput(const ConcreteDataMatcher& input, Lifetime lifetime) const
 {
   auto result = mPaths.find(input);
   if (result != mPaths.end()) {

--- a/Utilities/DataSampling/src/Dispatcher.cxx
+++ b/Utilities/DataSampling/src/Dispatcher.cxx
@@ -202,12 +202,11 @@ Inputs Dispatcher::getInputSpecs()
 
   // Add data inputs. Avoid duplicates and inputs which include others (e.g. "TST/DATA" includes "TST/DATA/1".
   for (const auto& policy : mPolicies) {
-    for (const auto& [potentiallyNewInput, _policyOutput] : policy->getPathMap()) {
-      (void)_policyOutput;
+    for (const auto& path : policy->getPathMap()) {
+      auto& potentiallyNewInput = path.first;
 
       // The idea is that we remove all existing inputs which are covered by the potentially new input.
       // If there are none which are broader than the new one, then we add it.
-      // I hope this is enough for all corner cases, but I am not 100% sure.
       auto newInputIsBroader = [&potentiallyNewInput](const InputSpec& other) {
         return DataSpecUtils::includes(potentiallyNewInput, other);
       };

--- a/Utilities/DataSampling/src/Dispatcher.cxx
+++ b/Utilities/DataSampling/src/Dispatcher.cxx
@@ -38,9 +38,6 @@ namespace o2::utilities
 Dispatcher::Dispatcher(std::string name, const std::string reconfigurationSource)
   : mName(name), mReconfigurationSource(reconfigurationSource)
 {
-  header::DataDescription timerDescription;
-  timerDescription.runtimeInit(("TIMER-" + name).substr(0, 16).c_str());
-  inputs.emplace_back(InputSpec{"timer-stats", "DS", timerDescription, 0, Lifetime::Timer});
 }
 
 Dispatcher::~Dispatcher() = default;
@@ -55,15 +52,16 @@ void Dispatcher::init(InitContext& ctx)
     std::unique_ptr<ConfigurationInterface> cfg = ConfigurationFactory::getConfiguration(mReconfigurationSource);
     policiesTree = cfg->getRecursive("dataSamplingPolicies");
     mPolicies.clear();
-  } else {
+  } else if (ctx.options().isSet("sampling-config-ptree")) {
     policiesTree = ctx.options().get<boost::property_tree::ptree>("sampling-config-ptree");
     mPolicies.clear();
-  }
+  } else
+    ; // we use policies declared during workflow init.
 
   for (auto&& policyConfig : policiesTree) {
     // we don't want the Dispatcher to exit due to one faulty Policy
     try {
-      mPolicies.emplace_back(std::make_shared<DataSamplingPolicy>(policyConfig.second));
+      mPolicies.emplace_back(std::make_shared<DataSamplingPolicy>(DataSamplingPolicy::fromConfiguration(policyConfig.second)));
     } catch (std::exception& ex) {
       LOG(WARN) << "Could not load the Data Sampling Policy '"
                 << policyConfig.second.get_optional<std::string>("id").value_or("") << "', because: " << ex.what();
@@ -188,22 +186,9 @@ void Dispatcher::sendFairMQ(FairMQDevice* device, const DataRef& inputData, cons
   int64_t bytesSent = device->Send(message, fairMQChannel);
 }
 
-void Dispatcher::registerPath(const std::pair<InputSpec, OutputSpec>& path)
+void Dispatcher::registerPolicy(std::unique_ptr<DataSamplingPolicy>&& policy)
 {
-  //todo: take care of inputs inclusive in others, when subSpec matchers are supported
-  auto cmp = [a = path.first](const InputSpec b) {
-    return a.matcher == b.matcher && a.lifetime == b.lifetime;
-  };
-
-  if (std::find_if(inputs.begin(), inputs.end(), cmp) == inputs.end()) {
-    inputs.push_back(path.first);
-    LOG(DEBUG) << "Registering input " << DataSpecUtils::describe(path.first);
-  } else {
-    LOG(DEBUG) << "Input " << DataSpecUtils::describe(path.first)
-               << " already registered";
-  }
-
-  outputs.push_back(path.second);
+  mPolicies.emplace_back(std::move(policy));
 }
 
 const std::string& Dispatcher::getName()
@@ -213,12 +198,66 @@ const std::string& Dispatcher::getName()
 
 Inputs Dispatcher::getInputSpecs()
 {
-  return inputs;
+  Inputs declaredInputs;
+
+  // Add data inputs. Avoid duplicates and inputs which include others (e.g. "TST/DATA" includes "TST/DATA/1".
+  for (const auto& policy : mPolicies) {
+    for (const auto& [potentiallyNewInput, _policyOutput] : policy->getPathMap()) {
+      (void)_policyOutput;
+
+      // The idea is that we remove all existing inputs which are covered by the potentially new input.
+      // If there are none which are broader than the new one, then we add it.
+      // I hope this is enough for all corner cases, but I am not 100% sure.
+      auto newInputIsBroader = [&potentiallyNewInput](const InputSpec& other) {
+        return DataSpecUtils::includes(potentiallyNewInput, other);
+      };
+      declaredInputs.erase(std::remove_if(declaredInputs.begin(), declaredInputs.end(), newInputIsBroader), declaredInputs.end());
+
+      auto declaredInputIsBroader = [&potentiallyNewInput](const InputSpec& other) {
+        return DataSpecUtils::includes(other, potentiallyNewInput);
+      };
+      if (std::none_of(declaredInputs.begin(), declaredInputs.end(), declaredInputIsBroader)) {
+        declaredInputs.push_back(potentiallyNewInput);
+      }
+    }
+  }
+
+  // add timer input
+  header::DataDescription timerDescription;
+  timerDescription.runtimeInit(("TIMER-" + mName).substr(0, 16).c_str());
+  declaredInputs.emplace_back(InputSpec{"timer-stats", "DS", timerDescription, 0, Lifetime::Timer});
+
+  return declaredInputs;
 }
 
 Outputs Dispatcher::getOutputSpecs()
 {
-  return outputs;
+  Outputs declaredOutputs;
+  for (const auto& policy : mPolicies) {
+    for (const auto& [_policyInput, policyOutput] : policy->getPathMap()) {
+      (void)_policyInput;
+      // In principle Data Sampling Policies should have different outputs.
+      // We may add a check to be very gentle with users.
+      declaredOutputs.push_back(policyOutput);
+    }
+  }
+  return declaredOutputs;
+}
+framework::Options Dispatcher::getOptions()
+{
+  o2::framework::Options options;
+  for (const auto& policy : mPolicies) {
+    if (!policy->getFairMQOutputChannel().empty()) {
+      if (!options.empty()) {
+        throw std::runtime_error("Maximum one policy with raw FairMQ channel is allowed, more have been declared.");
+      }
+      options.push_back({"channel-config", VariantType::String, policy->getFairMQOutputChannel().c_str(), {"Out-of-band channel config"}});
+      LOG(DEBUG) << " - registering output FairMQ channel '" << policy->getFairMQOutputChannel() << "'";
+    }
+  }
+  options.push_back({"period-timer-stats", framework::VariantType::Int, 10 * 1000000, {"Dispatcher's stats timer period"}});
+
+  return options;
 }
 
 } // namespace o2::utilities

--- a/Utilities/DataSampling/test/test_DataSampling.cxx
+++ b/Utilities/DataSampling/test/test_DataSampling.cxx
@@ -14,6 +14,8 @@
 #include <boost/test/unit_test.hpp>
 
 #include "DataSampling/DataSampling.h"
+#include "DataSampling/Dispatcher.h"
+#include "DataSampling/DataSamplingPolicy.h"
 #include "Framework/DataProcessingHeader.h"
 #include "Framework/ExternalFairMQDeviceProxy.h"
 #include "DataSampling/DataSamplingReadoutAdapter.h"
@@ -223,4 +225,79 @@ BOOST_AUTO_TEST_CASE(DataSamplingEmptyConfig)
 
   WorkflowSpec workflow;
   BOOST_CHECK_NO_THROW(DataSampling::GenerateInfrastructure(workflow, configFilePath));
+}
+
+BOOST_AUTO_TEST_CASE(DataSamplingOverlappingInputs)
+{
+  {
+    // policy3 includes 1 and 2, so we should have only one inputspec for data, one for timer
+    Dispatcher dispatcher("dispatcher", "");
+    auto policy1 = std::make_unique<DataSamplingPolicy>("policy1");
+    policy1->registerPath({"vcxz", "TST", "AAAA", 0}, {{"erwv"}, "DS", "AAAA"});
+
+    auto policy2 = std::make_unique<DataSamplingPolicy>("policy2");
+    policy2->registerPath({"fdsa", "TST", "AAAA", 0}, {{"fdsf"}, "DS", "BBBB"});
+
+    auto policy3 = std::make_unique<DataSamplingPolicy>("policy3");
+    policy3->registerPath({"asdf", {"TST", "AAAA"}}, {{"erwv"}, "DS", "CCCC"});
+
+    dispatcher.registerPolicy(std::move(policy1));
+    dispatcher.registerPolicy(std::move(policy2));
+    dispatcher.registerPolicy(std::move(policy3));
+
+    auto inputs = dispatcher.getInputSpecs();
+
+    BOOST_REQUIRE_EQUAL(inputs.size(), 2);
+    BOOST_CHECK_EQUAL(inputs[0], (InputSpec{"asdf", {"TST", "AAAA"}}));
+    BOOST_CHECK_EQUAL(inputs[1], (InputSpec{"timer-stats", "DS", "TIMER-dispatcher", 0, Lifetime::Timer}));
+  }
+
+  {
+    // policy3 includes 1 and 2, so we should have only one inputspec for data, one for timer
+    // same as before, but different order of registration
+    Dispatcher dispatcher("dispatcher", "");
+    auto policy1 = std::make_unique<DataSamplingPolicy>("policy1");
+    policy1->registerPath({"vcxz", "TST", "AAAA", 0}, {{"erwv"}, "DS", "AAAA"});
+
+    auto policy2 = std::make_unique<DataSamplingPolicy>("policy2");
+    policy2->registerPath({"fdsa", "TST", "AAAA", 0}, {{"fdsf"}, "DS", "BBBB"});
+
+    auto policy3 = std::make_unique<DataSamplingPolicy>("policy3");
+    policy3->registerPath({"asdf", {"TST", "AAAA"}}, {{"erwv"}, "DS", "CCCC"});
+
+    dispatcher.registerPolicy(std::move(policy3));
+    dispatcher.registerPolicy(std::move(policy1));
+    dispatcher.registerPolicy(std::move(policy2));
+
+    auto inputs = dispatcher.getInputSpecs();
+
+    BOOST_REQUIRE_EQUAL(inputs.size(), 2);
+    BOOST_CHECK_EQUAL(inputs[0], (InputSpec{"asdf", {"TST", "AAAA"}}));
+    BOOST_CHECK_EQUAL(inputs[1], (InputSpec{"timer-stats", "DS", "TIMER-dispatcher", 0, Lifetime::Timer}));
+  }
+
+  {
+    // three different inputs + timer
+    Dispatcher dispatcher("dispatcher", "");
+    auto policy1 = std::make_unique<DataSamplingPolicy>("policy1");
+    policy1->registerPath({"vcxz", "TST", "AAAA", 0}, {{"erwv"}, "DS", "AAAA"});
+
+    auto policy2 = std::make_unique<DataSamplingPolicy>("policy2");
+    policy2->registerPath({"sfsd", "TST", "BBBB", 0}, {{"fdsf"}, "DS", "BBBB"});
+
+    auto policy3 = std::make_unique<DataSamplingPolicy>("policy3");
+    policy3->registerPath({"asdf", {"TST", "CCCC"}}, {{"erwv"}, "DS", "CCCC"});
+
+    dispatcher.registerPolicy(std::move(policy1));
+    dispatcher.registerPolicy(std::move(policy2));
+    dispatcher.registerPolicy(std::move(policy3));
+
+    auto inputs = dispatcher.getInputSpecs();
+
+    BOOST_REQUIRE_EQUAL(inputs.size(), 4);
+    BOOST_CHECK_EQUAL(inputs[0], (InputSpec{"vcxz", "TST", "AAAA"}));
+    BOOST_CHECK_EQUAL(inputs[1], (InputSpec{"sfsd", "TST", "BBBB"}));
+    BOOST_CHECK_EQUAL(inputs[2], (InputSpec{"asdf", {"TST", "CCCC"}}));
+    BOOST_CHECK_EQUAL(inputs[3], (InputSpec{"timer-stats", "DS", "TIMER-dispatcher", 0, Lifetime::Timer}));
+  }
 }

--- a/Utilities/DataSampling/test/test_DataSamplingPolicy.cxx
+++ b/Utilities/DataSampling/test/test_DataSamplingPolicy.cxx
@@ -14,12 +14,14 @@
 #include <boost/test/unit_test.hpp>
 #include <boost/property_tree/ptree.hpp>
 
+#include "DataSampling/DataSamplingConditionFactory.h"
 #include "DataSampling/DataSamplingPolicy.h"
 #include "Framework/DataRef.h"
 #include "Framework/DataProcessingHeader.h"
 
 using namespace o2::framework;
 using namespace o2::utilities;
+using namespace o2::header;
 
 // an example of DataSamplingPolicy JSON object
 // {
@@ -29,19 +31,7 @@ using namespace o2::utilities;
 //     "aidlalala1",
 //     "aidlalala2"
 //   ]
-//   "dataHeaders" : [
-//     {
-//       "binding" : "clusters",
-//       "dataOrigin" : "TPC",
-//       "dataDescription" : "CLUSTERS"
-//     },
-//     {
-//       "binding" : "tracks",
-//       "dataOrigin" : "TPC",
-//       "dataDescription" : "TRACKS"
-//     }
-//   ],
-//   "subSpec" : "*",
+//   "query" : "c:TST/CHLEB/33;m:TST/MLEKO/33",
 //   "samplingConditions" : [
 //     {
 //       "condition" : "random",
@@ -52,10 +42,9 @@ using namespace o2::utilities;
 //   "blocking" : "false"
 // }
 
-BOOST_AUTO_TEST_CASE(DataSamplingPolicyConfiguration)
+BOOST_AUTO_TEST_CASE(DataSamplingPolicyFromConfiguration)
 {
   using boost::property_tree::ptree;
-  DataSamplingPolicy policy;
 
   ptree config;
   config.put("id", "my_policy");
@@ -70,28 +59,57 @@ BOOST_AUTO_TEST_CASE(DataSamplingPolicyConfiguration)
   config.add_child("samplingConditions", samplingConditions);
   config.put("blocking", "false");
 
-  policy.configure(config);
+  {
+    auto policy = std::move(DataSamplingPolicy::fromConfiguration(config));
 
-  BOOST_CHECK_EQUAL(policy.getName(), "my_policy");
-  BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "CHLEB", 33})) == (Output{"DS", "my_policy0", 33}));
-  BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "MLEKO", 33})) == (Output{"DS", "my_policy1", 33}));
-  const auto& map = policy.getPathMap();
-  BOOST_CHECK((*map.find(ConcreteDataMatcher{"TST", "CHLEB", 33})).second == (OutputSpec{"DS", "my_policy0", 33}));
-  BOOST_CHECK((*map.find(ConcreteDataMatcher{"TST", "MLEKO", 33})).second == (OutputSpec{"DS", "my_policy1", 33}));
-  BOOST_CHECK_EQUAL(map.size(), 2);
+    BOOST_CHECK_EQUAL(policy.getName(), "my_policy");
+    BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "CHLEB", 33})) == (Output{"DS", "my_policy0", 33}));
+    BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "MLEKO", 33})) == (Output{"DS", "my_policy1", 33}));
+    const auto& map = policy.getPathMap();
+    BOOST_CHECK((*map.find(ConcreteDataMatcher{"TST", "CHLEB", 33})).second == (OutputSpec{"DS", "my_policy0", 33}));
+    BOOST_CHECK((*map.find(ConcreteDataMatcher{"TST", "MLEKO", 33})).second == (OutputSpec{"DS", "my_policy1", 33}));
+    BOOST_CHECK_EQUAL(map.size(), 2);
 
-  BOOST_CHECK(policy.match(ConcreteDataMatcher{"TST", "CHLEB", 33}));
-  BOOST_CHECK(!policy.match(ConcreteDataMatcher{"TST", "SZYNKA", 33}));
+    BOOST_CHECK(policy.match(ConcreteDataMatcher{"TST", "CHLEB", 33}));
+    BOOST_CHECK(!policy.match(ConcreteDataMatcher{"TST", "SZYNKA", 33}));
 
-  DataProcessingHeader dph{555, 0};
-  o2::header::Stack headerStack{dph};
-  DataRef dr{nullptr, reinterpret_cast<const char*>(headerStack.data()), nullptr};
-  policy.decide(dr); // just make sure it does not crash
+    DataProcessingHeader dph{555, 0};
+    o2::header::Stack headerStack{dph};
+    DataRef dr{nullptr, reinterpret_cast<const char*>(headerStack.data()), nullptr};
+    policy.decide(dr); // just make sure it does not crash
+  }
 
   config.put("id", "too-long-policy-name");
-  policy.configure(config);
-  BOOST_CHECK_EQUAL(policy.getName(), "too-long-polic");
-  BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "CHLEB", 33})) == (Output{"DS", "too-long-polic0", 33}));
-  BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "MLEKO", 33})) == (Output{"DS", "too-long-polic1", 33}));
-  BOOST_CHECK_EQUAL(policy.getPathMap().size(), 2); // previous paths should be cleared
+
+  {
+    auto policy = std::move(DataSamplingPolicy::fromConfiguration(config));
+    BOOST_CHECK_EQUAL(policy.getName(), "too-long-policy-name");
+    BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "CHLEB", 33})) == (Output{"DS", "too-long-polic0", 33}));
+    BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "MLEKO", 33})) == (Output{"DS", "too-long-polic1", 33}));
+  }
+}
+
+BOOST_AUTO_TEST_CASE(DataSamplingPolicyFromMethods)
+{
+  DataSamplingPolicy policy("my_policy");
+  auto conditionNConsecutive = DataSamplingConditionFactory::create("nConsecutive");
+  BOOST_REQUIRE(conditionNConsecutive);
+
+  boost::property_tree::ptree config;
+  config.put("samplesNumber", 3);
+  config.put("cycleSize", 10);
+  conditionNConsecutive->configure(config);
+
+  policy.registerCondition(std::move(conditionNConsecutive));
+
+  policy.registerPath({"tststs", {"TST", "CHLEB"}}, {{"asdf"}, "AA", "BBBB"});
+  BOOST_CHECK((policy.prepareOutput(ConcreteDataMatcher{"TST", "CHLEB", 33})) == (Output{"AA", "BBBB", 33}));
+}
+
+BOOST_AUTO_TEST_CASE(DataSamplingPolicyStaticMethods)
+{
+  BOOST_CHECK(DataSamplingPolicy::createPolicyDataOrigin() == DataOrigin("DS"));
+  BOOST_CHECK(DataSamplingPolicy::createPolicyDataDescription("asdf", 0) == DataDescription("asdf0"));
+  BOOST_CHECK(DataSamplingPolicy::createPolicyDataDescription("asdfasdfasdfasdf", 0) == DataDescription("asdfasdfasdfas0"));
+  BOOST_CHECK(DataSamplingPolicy::createPolicyDataDescription("asdfasdfasdfasdf", 10) == DataDescription("asdfasdfasdfas10"));
 }

